syntax = "proto3";

option go_package = "";
package network;

message Query {
    int32 number = 1;
    string message = 2;
}

message QueryResponse {
    string response = 1;
}

message Node {
    int64 id = 1;
    map<string, string> fields = 2;    
}

message Edge {
    string label = 1;
    map<string, string> fields = 2;
}

message CreateNodeReq {
    int64 id = 1;
    map<string, string> fields = 2;
}

message CreateNodeResponse {
    Node node = 1;
}

message CreateEdgeReq {
    string label = 1;
    map<string, string> fields = 2;
}

message CreateEdgeResponse {
    Edge edge = 1;
}

message CreateRelationReq {
    Node source = 1;
    Node destination = 2;
    Edge edgeData = 3;
}

message CraeteRelationResponse {
    Node source = 1;
    Node destination = 2;
    Edge edge = 3;
}

message DeleteNodeReq {
    Node node = 1;
}

message DeleteNodeResponse {
    Node node = 1;
}

message DeleteEdgeReq {
    Edge edge = 1;
}

message DeleteEdgeResponse {
    Edge edge = 1;
}

service QueryService {
    rpc ExecuteQuery (Query) returns (QueryResponse) {};
    rpc CreateNode (CreateNodeReq) returns (CreateNodeResponse) {};
    rpc DeleteNode (DeleteNodeReq) returns (DeleteNodeResponse) {};
    rpc CreateRelation (CreateRelationReq) returns (CraeteRelationResponse) {};
    rpc CreateEdge (CreateEdgeReq) returns (CreateEdgeResponse) {};
    rpc DeleteEdge (DeleteEdgeReq) returns (DeleteEdgeResponse) {};
    // rpc SearchNodeById (Query) returns (QueryResponse) {};
    // rpc SearchNodeByCondition (Query) returns (QueryResponse) {};
    rpc SearchNode (Query) returns (QueryResponse) {};
}
